### PR TITLE
Prompt the last admin before letting them leave a room

### PR DESCRIPTION
**Pull request by ClaireGizard**
#345

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileViewModel.kt [hunk 1]
```diff
@@ -112,6 +117,39 @@ class RoomProfileViewModel @AssistedInject constructor(
...
+    }
+
+    private fun buildAdminMembersList(powerLevelsContent: PowerLevelsContent, roomMembers: List<RoomMemberSummary>): List<RoomMemberSummary> {
+        val powerLevelsHelper = PowerLevelsHelper(powerLevelsContent)
+
+        return roomMembers
+                .mapNotNull { roomMember ->
+                    val userRole = powerLevelsHelper.getUserRole(roomMember.userId)
+                    if (userRole == Role.Admin) roomMember else null
+                }
```
Comment (yostyle) [inline vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileViewModel.kt:150]:
```suggestion
        return roomMembers
                .mapNotNull { roomMember ->
                    if (PowerLevelsHelper(powerLevelsContent).getUserRole(roomMember.userId) == Role.Admin) roomMember else null
                }
```

**Review by yostyle — COMMENTED** [review]
LGTM just one comment

**Comment by yostyle** [comment]
SGTM
@Florian14 could you review this PR too ?

**Review by yostyle — APPROVED** [review]

File: vector/src/main/res/values-fr/strings_tchap.xml [hunk 2]
```diff
@@ -32,7 +32,8 @@
     <string name="tchap_dialog_prompt_new_direct_chat">Vous n\'avez pas de message direct avec %s, voulez-vous lui envoyer une invitation ?</string>
 
     <!-- tchap new strings -->
-    <string name="tchap_room_admin_leave_prompt_msg">Voulez-vous vraiment quitter cette conversation ?\n\nElle ne sera plus administrée, et vous risquez de ne plus pouvoir la joindre de nouveau.</string>
+    <string name="tchap_room_last_admin_leave_prompt_msg">Ce salon ne sera plus administrée, et vous risquez de ne plus pouvoir la joindre de nouveau.</string>
```
Comment (giomfo) [inline vector/src/main/res/values-fr/strings_tchap.xml:35]:
```suggestion
    <string name="tchap_room_last_admin_leave_prompt_msg">Ce salon ne sera plus administré, et vous risquez de ne plus pouvoir le joindre de nouveau.</string>
```

File: vector/src/main/res/values/strings_tchap.xml [hunk 3]
```diff
@@ -32,7 +32,8 @@
     <string name="tchap_dialog_prompt_new_direct_chat">You don\'t have a direct chat with %s, do you want to send him an invitation?</string>
 
     <!-- tchap new strings -->
-    <string name="tchap_room_admin_leave_prompt_msg">Are you sure you want to leave the room?\n\nThe room will no longer be administered, and you may not be able to join it again</string>
+    <string name="tchap_room_admin_leave_prompt_msg">This room is not public. You will not able to join it without an invitation.</string>
```
Comment (giomfo) [inline vector/src/main/res/values/strings_tchap.xml:35]:
```suggestion
    <string name="tchap_room_admin_leave_prompt_msg">This room is not public. You will not be able to join it without an invitation.</string>
```

File: vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileFragment.kt [hunk 4]
```diff
@@ -271,13 +271,18 @@ class RoomProfileFragment @Inject constructor(
...
-    override fun onLeaveRoomClicked() {
+    override fun onLeaveRoomClicked(isLastAdmin: Boolean) {
         val isPublicRoom = roomProfileViewModel.isPublicRoom()
         val message = buildString {
             append(getString(R.string.room_participants_leave_prompt_msg))
-            if (!isPublicRoom) {
+            if (!isLastAdmin) {
+                if (!isPublicRoom) {
+                    append("\n\n")
+                    append(getString(R.string.tchap_room_admin_leave_prompt_msg))
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileFragment.kt:281]: why don't you use the element string `room_participants_leave_private_warning`?

File: vector/src/main/res/values-fr/strings_tchap.xml [hunk 5]
```diff
@@ -32,7 +32,8 @@
     <string name="tchap_dialog_prompt_new_direct_chat">Vous n\'avez pas de message direct avec %s, voulez-vous lui envoyer une invitation ?</string>
 
     <!-- tchap new strings -->
-    <string name="tchap_room_admin_leave_prompt_msg">Voulez-vous vraiment quitter cette conversation ?\n\nElle ne sera plus administrée, et vous risquez de ne plus pouvoir la joindre de nouveau.</string>
+    <string name="tchap_room_last_admin_leave_prompt_msg">Ce salon ne sera plus administré, et vous risquez de ne plus pouvoir le joindre de nouveau.</string>
+    <string name="tchap_room_admin_leave_prompt_msg">Ce salon n\'est pas public. Vous ne pourrez pas le rejoindre sans invitation</string>
```
[the remark this answers is not in the document]
Comment (ClaireGizard) [reply vector/src/main/res/values-fr/strings_tchap.xml]: Good to know. I just deleted this string to use Tchap string

File: vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileViewModel.kt [hunk 6]
```diff
@@ -112,6 +117,36 @@ class RoomProfileViewModel @AssistedInject constructor(
...
+                                roomMember.userId == session.myUserId
+                            } != null)
+                }
+    }
+
+    private fun buildAdminMembersList(powerLevelsContent: PowerLevelsContent, roomMembers: List<RoomMemberSummary>): List<RoomMemberSummary> {
+        return roomMembers
+                .mapNotNull { roomMember ->
+                    if (PowerLevelsHelper(powerLevelsContent).getUserRole(roomMember.userId) == Role.Admin) roomMember else null
+                }
```
Comment (Florian14) [inline vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileViewModel.kt:147]: I think the `filter` operator does the trick:

```suggestion
        return roomMembers.filter { PowerLevelsHelper(powerLevelsContent).getUserRole(it.userId) == Role.Admin) }
```

File: vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileViewModel.kt [hunk 7]
```diff
@@ -112,6 +117,36 @@ class RoomProfileViewModel @AssistedInject constructor(
...
+                        .unwrap()
+        ) { roomMembers, powerLevelsContent ->
+            buildAdminMembersList(powerLevelsContent, roomMembers)
+        }
+                .execute { async ->
+                    copy(isLastAdmin = async.invoke()?.size == 1 &&
+                            async.invoke()?.find { roomMember ->
+                                roomMember.userId == session.myUserId
+                            } != null)
+                }
```
Comment (Florian14) [inline vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileViewModel.kt:140]: Not sure about the difference between execute and onEach but I suggest you these changes.
- `SingleOrNull` does the trick to check if there is a single element and return it if any.
- `OnEach` will be triggered once Flows are updated and do not return an `async` with a nullable value when invoking it.

```suggestion
                .distinctUntilChanged()
                .onEach { adminsMembers ->
                    setState {
                        copy(isLastAdmin = adminsMembers.singleOrNull()?.userId == session.myUserId)
                    }
                }
                .launchIn(viewModelScope)
```

**Comment by giomfo** [comment]
one more comment https://github.com/tchapgouv/tchap-android-v2/issues/345#issuecomment-1016219963

File: vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileFragment.kt [hunk 8]
```diff
@@ -271,13 +271,18 @@ class RoomProfileFragment @Inject constructor(
         ShortcutManagerCompat.requestPinShortcut(requireContext(), onShortcutReady.shortcutInfo, null)
     }
 
-    override fun onLeaveRoomClicked() {
+    override fun onLeaveRoomClicked(isLastAdmin: Boolean) {
         val isPublicRoom = roomProfileViewModel.isPublicRoom()
         val message = buildString {
             append(getString(R.string.room_participants_leave_prompt_msg))
-            if (!isPublicRoom) {
+            if (!isLastAdmin) {
```
Comment (Florian14) [inline vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileFragment.kt:278]: please add a `// Tchap:` comment

File: vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileViewModel.kt [hunk 9]
```diff
@@ -112,6 +118,33 @@ class RoomProfileViewModel @AssistedInject constructor(
                 }
     }
 
+    private fun observeAdminMembers() {
```
Comment (Florian14) [inline vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileViewModel.kt:121]: please add a `// Tchap:` comment before

**Review by Florian14 — COMMENTED** [review]
LGTM, just some small feedbacks about rebase

**Review by Florian14 — APPROVED** [review]
Thanks for the update!

**Review by giomfo — APPROVED** [review]
LGTM

**Pull request merged**